contracts/epochs/FlowEpoch.cdc

```diff
@@ -151,6 +151,77 @@ access(all) contract FlowEpoch {
         dkgIdMapping: {String: Int},
     )
 
+    /// The EpochRecover service event is emitted via the recoverEpoch governance transaction to execute an epoch recovery.
+    /// The epoch recovery process is used if the Epoch Preparation Protocol fails for any reason.
+    /// When this happens, the network enters Epoch Fallback Mode [EFM], wherein the Protocol State
+    /// extends the current epoch until an epoch recovery can take place. In general, while in EFM, the
+    /// Protocol State and the FlowEpoch smart contract may have conflicting views about the current
+    /// valid epoch state (epoch counter and phase). The epoch recovery process resolves this inconsistency
+    /// by injecting a new common shared state to proceed forward from. 
+    /// Concretely, the epoch recovery process fully specifies the parameters for a new epoch to transition into
+    /// (essentially all fields of the EpochSetup and EpochCommit events), called the recovery epoch.
+    /// The FlowEpoch smart contracts inserts the recovery epoch, starts it, and emits the EpochRecover event,
+    /// which contains the same data. When the EpochRecover event is processed by the protocol state, 
+    /// the Protocol State is updated to include the recovery epoch as well.
+    access(all) event EpochRecover (
+        /// The counter for the RecoveryEpoch.
+        /// This must be 1 greater than the current epoch counter, as reported by the Protocol State,
+        /// otherwise the EpochRecover event will be rejected (recovery process will fail).
+        counter: UInt64,
+
+        /// Identity table for the upcoming epoch with all node information.
+        /// including nodeID, staking key, networking key, networking address, role,
+        /// staking information, weight, and more.
+        nodeInfo: [FlowIDTableStaking.NodeInfo],
+
+        /// The first view (inclusive) of the RecoveryEpoch. This must be 1 greater
+        /// than the current epoch's final view, as reported by the Protocol State,
```

Comment:
In EFM, the Protocol State periodically adds extensions to the epoch, which increases the `FinalView` field for the epoch. The comment means that this `firstView` value in the service event must be exactly one greater than the current `FinalView` value, when the service event is processed. 

The `firstView` value comes from the `recoverEpoch` transaction arguments, not from the current view when the transaction is processed. So you can submit the transaction at any point, as long as the arguments contain the correct `firstView` value. 

With `resetEpoch`, we do a spork, then later submit a `resetEpoch` to synchronize the smart contract state with the Protocol State. The smart contract does the epoch transition "late"; before the `resetEpoch`, it has outdated epoch state. The way I've been thinking of this is like `resetEpoch`, but the smart contract does the epoch transition early instead of late. If the current view is 90, and the `firstView` is 100, the smart contract will do the transition right away, then the Protocol State will transition at view 100. Both are synchronized at view 100.

Reply:
okay, that makes sense